instrumentation/opentelemetry-instrumentation-openai/src/opentelemetry/instrumentation/openai/patch.py

```diff
@@ -0,0 +1,308 @@
+                )
+        set_event_completion(span, responses)
+
+    if getattr(result, "id", None):
+        set_span_attribute(span, GenAIAttributes.GEN_AI_RESPONSE_ID, result.id)
+
+    # Get the usage
+    if getattr(result, "usage", None):
+        set_span_attribute(
+            span,
+            GenAIAttributes.GEN_AI_USAGE_INPUT_TOKENS,
+            result.usage.prompt_tokens,
+        )
+        set_span_attribute(
+            span,
+            GenAIAttributes.GEN_AI_USAGE_OUTPUT_TOKENS,
+            result.usage.completion_tokens,
+        )
+        set_span_attribute(
+            span,
+            "gen_ai.usage.total_tokens",
+            result.usage.total_tokens,
+        )
+
+
+class StreamWrapper:
+    span: Span
+    response_id: str = ""
+    response_model: str = ""
+
+    def __init__(
+        self,
+        stream,
+        span,
+        prompt_tokens=0,
+        function_call=False,
+        tool_calls=False,
+    ):
+        self.stream = stream
+        self.span = span
+        self.prompt_tokens = prompt_tokens
+        self.function_call = function_call
+        self.tool_calls = tool_calls
+        self.result_content = []
+        self.completion_tokens = 0
+        self._span_started = False
+        self.setup()
+
+    def setup(self):
+        if not self._span_started:
+            self._span_started = True
+
+    def cleanup(self):
+        if self._span_started:
+            if self.response_model:
+                set_span_attribute(
+                    self.span,
+                    GenAIAttributes.GEN_AI_RESPONSE_MODEL,
+                    self.response_model,
+                )
+
+            if self.response_id:
+                set_span_attribute(
+                    self.span,
+                    GenAIAttributes.GEN_AI_RESPONSE_ID,
+                    self.response_id,
+                )
+
+            set_span_attribute(
+                self.span,
+                GenAIAttributes.GEN_AI_USAGE_INPUT_TOKENS,
+                self.prompt_tokens,
+            )
+            set_span_attribute(
+                self.span,
+                GenAIAttributes.GEN_AI_USAGE_OUTPUT_TOKENS,
+                self.completion_tokens,
+            )
+            set_span_attribute(
+                self.span,
+                "gen_ai.usage.total_tokens",
+                self.prompt_tokens + self.completion_tokens,
+            )
+            set_event_completion(
+                self.span,
+                [
+                    {
+                        "role": "assistant",
+                        "content": "".join(self.result_content),
+                    }
+                ],
+            )
+
+            self.span.set_status(StatusCode.OK)
+            self.span.end()
+            self._span_started = False
+
+    def __enter__(self):
+        self.setup()
+        return self
+
+    def __exit__(self, exc_type, exc_val, exc_tb):
+        self.cleanup()
```

Comment:
what happens if connection drops in the middle of the stream?

would error be reported here? We need to pass it and set status and error.type attribute based on it.